Get user input

shahn95 opened the pull request:
Accept user input. Will add input checking/validation at a later stage.

shahn95 committed:
Update .gitignore to include .calva

shahn95 committed:
Get input from user and store in info-map

Added functionality to accept 3 lines of input data.
Data is then decomposed and stored in a map: info-map.

yudistrange commented on mars-rover-new/src/mars_rover_new/core.clj:
```diff
@@ -1,7 +1,94 @@
+  (println "Welcome to Mars!")
+  (prn (user-input)))
+
+;; repl testing
+;; (add-plateau-size "10 100" info-map)
+;; (add-initial-position "0 0 N" info-map)
+;; (add-cmd-sequence "LLRRMM" info-map)
+
+;; (add-cmd-sequence "LLRRMM" (add-initial-position "0 0 N" (add-plateau-size "10 10" info-map)))
+
+
+;; (defn add-plateau-size
+;;   "Add plateau-size data to info-map"
+;;   [plateau-size, info-map]
+;;   (try
+;;     (let [[x y] (map #(Integer/parseInt %) (string/split plateau-size #" "))]
+;;       (assoc-in info-map [:plateau-size] {:x x
+;;                                           :y y}))
+;;     (catch Exception e
+;;       (println "Error processing plateau-size input, please try again.")
+;;       (println (str "(" e ")"))
+;;       (user-input))))
+
+;; (try
+;;   (Integer/parseInt (read-line))
+;;   (catch Exception e (str "caught exception: " (.getMessage e))))
```
Please don't check-in commented code. It adds noise to the PR.

shhivam replied:
+1

shahn95 replied:
Done!

yudistrange commented on mars-rover-new/src/mars_rover_new/core.clj:
```diff
@@ -1,7 +1,94 @@
+   (if (not-empty prompt)
+     (println prompt))
+   (let [input (string/trim (read-line))]
+     (string/upper-case input))))
```
Couple of things off with this fn.
- use `when`, when you don't need the `else` part of an `if` 
- fns should ideally be one s-exp. This fn works because the `if` block has an impure fn (println) in it. Functions in clojure return whatever is the result of the result of the last s-exp.

shahn95 replied:
Yes, started using `when` instead of `if` when I don't need an `else`

shahn95 replied:
@yudistrange Should I go ahead and resolve this?

yudistrange reviewed (commented):
Left some minor comments. The code looks alright otherwise.

Another general minor thing: Maybe setup a formatter. It will make the maps / bindings etc line up nicely.

shhivam commented on mars-rover-new/src/mars_rover_new/core.clj:
```diff
@@ -1,7 +1,94 @@
+
+;; input validation: check for number of args, catch errors during type conversion
+
+(declare add-plateau-size, add-rover-position, add-cmd-sequence)
```
You won't have to `declare` these non-binded vars if you move the `user-input` function at the bottom. 

In my experience, the function which ties a lot of functions together are written at the bottom so that all the downstream functions are declared before that larger function is called.

shahn95 replied:
Makes sense, done!

shhivam commented on mars-rover-new/src/mars_rover_new/core.clj:
```diff
@@ -1,7 +1,94 @@
+
+(declare add-plateau-size, add-rover-position, add-cmd-sequence)
+
+(def info-map {:plateau-size {:x nil
```
Consider renaming this. It's not very clear how it's being used at a first glance.

shahn95 replied:
This is a map to store plateau size, rover position, and command sequence. All of these come from the user. 

So this map has plateau size, which is info about Mars, and rover position + commands, which is info about the rover.

What do you think of:
user-data
mars-rover-data

shahn95 replied:
Finally changed to (default)-mars-rover-data

shhivam commented on mars-rover-new/src/mars_rover_new/core.clj:
```diff
@@ -1,7 +1,94 @@
+  "Add plateau-size data to info-map"
+  [plateau-size, info-map]
+  (let [[x y] (map #(Integer/parseInt %) (string/split plateau-size #" "))]
+    (assoc-in info-map [:plateau-size] {:x x
```
Consider using `assoc` instead of `assoc-in` since there is no modification of nested keys yet.

Same for `add-rover-position` and `add-cmd-sequence`

shahn95 replied:
edit: done!

shhivam commented on mars-rover-new/src/mars_rover_new/core.clj:
```diff
@@ -1,7 +1,64 @@
+  (let [input (string/trim (read-line))]
+    (string/upper-case input)))
+
+(defn user-input
```
[NON BLOCKER]
Can we think of better names for this function?

shahn95 replied:
Changed to process-user-input!

shhivam commented on mars-rover-new/src/mars_rover_new/core.clj:
```diff
@@ -15,20 +17,31 @@
-        rover-position (get-input)
-        cmd-sequence (get-input)]
-    (add-cmd-sequence cmd-sequence (add-rover-position rover-position (add-plateau-size plateau-size info-map)))))
+  [] 
+  (let [plateau-size (get-input "Enter plateau size:")
+        rover-position (get-input "Enter initial rover position:")
+        cmd-sequence (get-input "Enter command sequence:")]
+    (->> info-map
+         (add-plateau-size plateau-size)
+         (add-rover-position rover-position)
+         (add-cmd-sequence cmd-sequence))))
```
[GIT HYGIENE]
Was going through the PR commit-by-commit. This refactor doesn't really belong in `Add ability to pass user prompts to get-input`. 
Let's pull this change out into a new commit?

shhivam reviewed (changes requested):
- The logic looks good!
- Overall good job on the commit messages and atomicity.

Something for later:
Maybe we should add some documentation on why there is a new mars rover folder and why we deprecated the old mars rover.
If possible, could we rename this new project and remove the old one down the line?

shahn95 committed:
Add ability to include prompts for user input

shahn95 committed:
Refactor user-input function to use threading macro

shahn95 committed:
Remove commented code

shahn95 committed:
Use when instead of if in get-input

shahn95 committed:
Add input verification functions along with helper functions and sets

Added input verification functions for plateau-size, rover-position and
cmd-sequence. However, these are not currently used.
Also added helper functions and sets to assist verification functions.

shahn95 committed:
Improve str->int function to return result or nil if error

shahn95 committed:
Rename verify-plateau-size to valid-plateau-size? and simplify logic

shahn95 committed:
Rename verify-rover-position to valid-rover-position? and simplify logic

shahn95 committed:
Rename verify-cmd-sequence to valid-cmd-sequence? and simplify logic

shahn95 committed:
Delete unused function string-seq-only-digits?

shahn95 committed:
Move user-input function towards the end of program

shahn95 committed:
Remove unused string-only-digits? function

shahn95 committed:
Change assoc-in to assoc in all functions

shahn95 committed:
Update names info-map -> mars-rover-data; user-input -> process-user-input

Also, first declaration of mars-rover-data is called default-mars-rover-data.

shahn95 committed:
Remove unnecessary declare statement

shahn95 committed:
Simplify valid-cmd-sequence? logic

shahn95 committed:
Add calls to verification functions and error handling for invalid inputs

Using :pre to perform input validation in add-plateau-size, add-rover-position,
and add-cmd-sequence. Assertion errors are caught in process-user-input. On error,
control is returned to the beginning of process-user-input.

shhivam commented on mars-rover-new/src/mars_rover_new/core.clj:
```diff
@@ -1,7 +1,118 @@
+(defn in-range?
+  "Check if val is within bounds (inclusive)"
+  [val lower-bound upper-bound]
+  (if (and (>= val lower-bound) (<= val upper-bound))
```
[NON-BLOCKING]
@shahn95 `<=` can work with more than 2 args.

So `(and (>= val lower-bound) (<= val upper-bound))` can be compressed into `(<= lower-bound val upper-bound)`

shhivam reviewed (approved):
We need to write tests for this but overall LGTM.
